alf/algorithms/sarsa_algorithm.py

```diff
@@ -13,41 +13,136 @@
-        'critic', 'target_critic'
+        'prev_observation', 'prev_step_type', 'actor', 'critics',
+        'target_critics'
     ],
     default_value=())
 SarsaInfo = namedtuple(
-    'SarsaInfo', ['action_distribution', 'actor_loss', 'critic', 'returns'])
-SarsaLossInfo = namedtuple(
-    'SarsaLossInfo', ['actor', 'critic'], default_value=())
+    'SarsaInfo', [
+        'action_distribution', 'actor_loss', 'alpha_loss', 'critics',
+        'target_critics', 'entropy_reward'
+    ],
+    default_value=())
+SarsaLossInfo = namedtuple('SarsaLossInfo', ['actor', 'critic', 'alpha'])
+
+nest_map = alf.nest.map_structure
+
+
+class FastCriticBias(nn.Module):
+    """
+    The motivation of this is to prevent the critic value becomes greater and
+    greater as training goes on because of the max operation.
+
+    FastCriticBias estimates expected return `b` by minimizing the following
+    expectation:
+        min_b E(b - r - discount * b)^2
+    where r is reward. `b` can be analytically solved as:
+        b = E(r) / (1 - E(discount))
+
+    We explicitly enforce that E(critic - b) = 0 by estimating E(critic) and
+    use b - E(critic) as bias to offset critic.
+
+    Pseudocode of using it:
+        critic = critic_network(...)
+        fast_bias.update(critic, reward, discount, mask)
+        new_critic = critic + fast_bias.get()
+        # use new_critic for training
+    """
+
+    def __init__(self, num_critics=1, speed=10):
+        """Create an instance of FastCriticBias.
+
+        Args:
+            num_critics (int): number of critic functions to process
+            speeed (float): how fast to update the averages
+            name (str): name of this object
+        """
+        super().__init__()
+        spec = alf.TensorSpec((), torch.float32)
+        self._averager = AdaptiveAverager(
+            tensor_spec=[[spec] * num_critics, spec, spec], speed=speed)
+
+        self._biases = []
+        for i in range(num_critics):
+            buf = torch.tensor(0.)
+            self._biases.append(buf)
+            self.register_buffer("_bias%d" % i, buf)
+
+    def update(self, critics, reward, discount, mask=None):
+        """Update internal statistics.
+
+        Args:
+            critics (Tensor|list[Tensor]): critics to process.
+            reward (Tensor): reward received for the current step
+            discount (Tensor): discount for the future steps
+            mask (bool Tensor): If provided, only use the entries whose
+                corresponding element in `mask` is True.
+        """
+        if torch.any(mask):
+            if isinstance(critics, torch.Tensor):
+                critics = [critics]
+            critics, reward, discount = math_ops.weighted_reduce_mean(
+                [critics, reward, discount], mask)
+            self._averager.average([critics, reward, discount])
+
+            critics, reward, discount = self._averager.get()
```

Comment:
These two lines can be combined into one because `average` by default will return.

Reply:
Fixed.